### ISSUE TITLE
New plugin: False positive report

### ISSUE THREAD

**Issue by azurit** [coreruleset/coreruleset#2486]
### Motivation

Handling of false positives is a hard task and cannot be done without proper tools. This plugin can be used to do most of the work.

### Proposed solution

Main purpose of the plugin is to send a notification for every request which seems to be blocked based on false positive alarms, togather with all information needed to write an exclusion rule. These filters can be used to set what should be considered as FP:
 - GeoIP
 - filter by rule ID
 - filter by rule message
 - filter by REQUEST_URI

Notifications are send by e-mail.

Plugin is able to suggest a simple exclusion rules which can be directly used to prevent blocking. Example of a notification is below.

Plugin is ready and was used in production for few months.

Example of a notification:
```
Hi, this is OWASP ModSecurity Core Rule Set False Positive Report Plugin!

According to your configuration, blocking of this request may be based on false positive alarms.

=== BASIC DATA ===
DATETIME: 2022-04-09 18:09:09
UNIQUE_ID: YlGvpb3DXsr7Xa79lav3BgAAAD4

REQUEST_LINE: GET /?a=/etc/passwd HTTP/1.1
SERVER_NAME: example.com
REMOTE_ADDR: 1.2.3.4

RESPONSE_STATUS: 403

=== HEADERS ===
Host: example.com
User-Agent: Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:99.0) Gecko/20100101 Firefox/99.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8
Accept-Language: sk,cs;q=0.8,en-US;q=0.5,en;q=0.3
Accept-Encoding: gzip, deflate
DNT: 1
Connection: keep-alive
Upgrade-Insecure-Requests: 1

=== POST ARGUMENTS ===


=== RAW REQUEST BODY ===


=== EXCLUSION RULE SUGGESTION ===
SecRule REQUEST_FILENAME "@endsWith /" \
    "id:<fix-me>,\
    phase:1,\
    pass,\
    t:none,\
    nolog,\
    ctl:ruleRemoveTargetById=930120;ARGS:a"

=== FILTERED LOGS ===
[file "apache2_util.c"] [line 269] [level 3] [client 1.2.3.4] ModSecurity: Access denied with code 403 (phase 2). Matched phrase "etc/passwd" at ARGS:a. [file "/usr/share/modsecurity-crs-new/rules/REQUEST-930-APPLICATION-ATTACK-LFI.conf"] [line "97"] [id "930120"] [msg "OS File Access Attempt"] [data "Matched Data: etc/passwd found within ARGS:a: /etc/passwd"] [severity "CRITICAL"] [ver "OWASP_CRS/3.3.2"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-lfi"] [tag "paranoia-level/1"] [tag "OWASP_CRS"] [tag "capec/1000/255/153/126"] [tag "PCI/6.5.4"] [hostname "example.com"] [uri "/"] [unique_id "YlGvpb3DXsr7Xa79lav3BgAAAD4"]

=== ALL LOGS ===
[file "apache2_util.c"] [line 269] [level 3] [client 1.2.3.4] ModSecurity: Access denied with code 403 (phase 2). Matched phrase "etc/passwd" at ARGS:a. [file "/usr/share/modsecurity-crs-new/rules/REQUEST-930-APPLICATION-ATTACK-LFI.conf"] [line "97"] [id "930120"] [msg "OS File Access Attempt"] [data "Matched Data: etc/passwd found within ARGS:a: /etc/passwd"] [severity "CRITICAL"] [ver "OWASP_CRS/3.3.2"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-lfi"] [tag "paranoia-level/1"] [tag "OWASP_CRS"] [tag "capec/1000/255/153/126"] [tag "PCI/6.5.4"] [hostname "example.com"] [uri "/"] [unique_id "YlGvpb3DXsr7Xa79lav3BgAAAD4"]

=== FULL REQUEST ===
GET /?a=/etc/passwd HTTP/1.1

Host: example.com
User-Agent: Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:99.0) Gecko/20100101 Firefox/99.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8
Accept-Language: sk,cs;q=0.8,en-US;q=0.5,en;q=0.3
Accept-Encoding: gzip, deflate
DNT: 1
Connection: keep-alive
Upgrade-Insecure-Requests: 1



=== RESPONSE BODY ===
Response body is retrieved only if at least one rule matches it.
```

**Comment by azurit** [comment]
More complex example of suggested exclusion rule:
```
$ curl "http://example.com/a.sql?a=<html>" -H "Content-Type: application/reports+json"

...
=== EXCLUSION RULE SUGGESTION ===
SecRule REQUEST_FILENAME "@endsWith /a.sql" \
    "id:<fix-me>,\
    phase:1,\
    pass,\
    t:none,\
    nolog,\
    setvar:'tx.allowed_request_content_type=%{tx.allowed_request_content_type} |application/reports+json|',\
    setvar:'tx.restricted_extensions=%{tx.restricted_extensions} .sql/',\
    ctl:ruleRemoveTargetById=921130;ARGS:a,\
    ctl:ruleRemoveTargetById=941320;ARGS:a"

...
```

**Comment by dune73** [comment]
Hey @azurit, how do you feel about this right now? Still interested in making this official? Still using it.

I'd rather not be part of a discussion around this, since I have different scripts published that are more flexible / versatile, but apparently do not work as a plugin, but based on the logs. So I have different interests here and will keep out.

**Comment by azurit** [comment]
@dune73 I'm still interested. Just completed the [XXE plugins](https://github.com/coreruleset/coreruleset/issues/2502#issuecomment-1210550761) so now i will look at this.

**Comment by azurit** [comment]
Here is a new plugin - False Positive Report:
https://github.com/azurit/modsecurity-false-positive-report-plugin

**Comment by Xhoenix** [comment]
Since the plugin has been released and been receiving positive feedback, should we close this issue?

**Comment by azurit** [comment]
No until it's decided if it's going to be accepted as official plugin or not.

**Comment by EsadCetiner** [comment]
@azurit Does the plugin have any way to filter out true positives to avoid spamming the user with false positives reports when they are in fact an attack?

@theseion @fzipi Thoughts, should we accept this as an official plugin?

**Comment by azurit** [comment]
> [@azurit](https://github.com/azurit) Does the plugin have any way to filter out true positives to avoid spamming the user with false positives reports when they are in fact an attack?

[Yes](https://github.com/azurit/modsecurity-false-positive-report-plugin?tab=readme-ov-file#filters).

**Comment by Xhoenix** [comment]
I came across an issue yesterday. I setup an OpenVAS scan and there were thousands of requests within a few hours. Although the scan stopped last night, I was still getting mails until today morning 5 AM. 

Could be the plugin isn't multi-threaded and able to handle too many requests simultaneously?

**Comment by azurit** [comment]
@Xhoenix Bottleneck wasn't the plugin itself or a web server but SMTP server - all those messages were probably waiting in SMTP queue.

**Comment by theseion** [comment]
We've decided to accept the plugin under the condition that the documentation makes it clear that we cannot support users with their SMTP setup and that users have to have a good handle on their FPs to avoid being flooded with e-mails.
